### PR TITLE
escrow: Migrate to cosmwasm v0.11.0

### PR DESCRIPTION
**Pull request by orkunkl**
Related to https://github.com/CosmWasm/cosmwasm-examples/issues/94
Don't merge until cosmwasm/v0.11.0 is released

### CODE REVIEW

File: escrow/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+use cosmwasm_std::StdError;
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum ContractError {
+    #[error("{0}")]
+    Std(#[from] StdError),
+
+    #[error("unauthorized")]
```
Comment (maurolacy) [inline escrow/src/error.rs:9]: This is a matter of taste, but I prefer capitalization of error messages, i. e. `"Unauthorized"`.

File: escrow/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    Unauthorized {},
+
+    #[error("escrow expired (end_height {end_height})")]
+    EscrowExpiredHeight { end_height: u64 },
```
Comment (maurolacy) [inline escrow/src/error.rs:13]: Again a matter of taste, but I think `Escrow` in the error enum is redundant. I would just use `ExpiredHeight`, and so on.

File: escrow/src/contract.rs [hunk 3]
```diff
@@ -1,122 +1,132 @@
...
+        source: deps.api.canonical_address(&info.sender)?,
         end_height: msg.end_height,
         end_time: msg.end_time,
     };
-    if state.is_expired(&env) {
-        Err(StdError::generic_err("creating expired escrow"))
-    } else {
-        config(&mut deps.storage).save(&state)?;
-        Ok(InitResponse::default())
+    match state.is_expired(&env) {
```
Comment (maurolacy) [inline escrow/src/contract.rs:23]: Another matter of taste, but I like the `if` better. I think the intention is clear.

Also, I try to avoid doing structural changes when migrating; unless strictly necessary, fixing something, etc.

File: escrow/src/contract.rs [hunk 4]
```diff
@@ -1,122 +1,132 @@
...
 
-        send_tokens(
-            &deps.api,
-            &deps.api.canonical_address(&env.contract.address)?,
-            &state.recipient,
-            amount,
-            "approve",
-        )
+    // throws error if state is expired
+    match state.is_expired(&env) {
```
Comment (maurolacy) [inline escrow/src/contract.rs:57]: It's you call, but, same here.

These changes also hinder reviewing, as I need to pay attention / verify stuff that is not strictly related to the migration. You can also inadvertently introduce subtle bugs, etc.

Now, for me you can leave these changes (if you're sure they are correct). I'm just mentioning these things as general guidelines.

File: escrow/src/contract.rs [hunk 5]
```diff
@@ -1,122 +1,132 @@
...
-        // https://github.com/CosmWasm/wasmd/blob/master/x/wasm/internal/keeper/keeper.go#L185-L192
-        let balance = deps.querier.query_all_balances(&env.contract.address)?;
-        send_tokens(
-            &deps.api,
-            &deps.api.canonical_address(&env.contract.address)?,
-            &state.source,
-            balance,
-            "refund",
-        )
+    let err = state.is_expired(&env);
```
Comment (maurolacy) [inline escrow/src/contract.rs:87]: I can't help but mention again that the original version looks clearer to me.

File: escrow/src/contract.rs [hunk 6]
```diff
@@ -1,122 +1,132 @@
...
 }
 
 // this is a helper to move the tokens, so the business logic is easy to read
-fn send_tokens<A: Api>(
-    api: &A,
-    from_address: &CanonicalAddr,
-    to_address: &CanonicalAddr,
+fn send_tokens(
+    from_address: HumanAddr,
+    to_address: HumanAddr,
```
Comment (maurolacy) [inline escrow/src/contract.rs:109]: Nice.

File: escrow/src/contract.rs [hunk 7]
```diff
@@ -147,20 +157,33 @@ mod tests {
...
+                height,
+                time,
+                ..Default::default()
+            },
+            contract: ContractInfo {
+                address: HumanAddr::from(MOCK_CONTRACT_ADDR),
+            },
+        };
+        let info = mock_info(signer, sent);
+        return (env, info);
```
Comment (maurolacy) [inline escrow/src/contract.rs:177]: As there are no dependencies between `mock_env_height()` and `mock_info()`, there's no real need to handle them together.

File: escrow/src/contract.rs [hunk 8]
```diff
@@ -188,20 +211,20 @@ mod tests {
...
+        let mut deps = mock_dependencies(&[]);
 
         let msg = init_msg_expire_by_height(1000);
-        let env = mock_env_height("creator", &coins(1000, "earth"), 1001, 0);
-        let res = init(&mut deps, env, msg);
+        let (env, info) = mock_env_info_height("creator", &coins(1000, "earth"), 1001, 0);
+        let res = init(&mut deps, env, info, msg);
         match res.unwrap_err() {
-            StdError::GenericErr { msg, .. } => assert_eq!(msg, "creating expired escrow"),
+            ContractError::EscrowExpiredHeight { .. } => {}
```
Comment (maurolacy) [inline escrow/src/contract.rs:220]: Nice! :-)

File: escrow/src/state.rs [hunk 9]
```diff
@@ -15,20 +16,23 @@ pub struct State {
 }
 
 impl State {
-    pub fn is_expired(&self, env: &Env) -> bool {
+    pub fn is_expired(&self, env: &Env) -> ContractError {
```
Comment (maurolacy) [inline escrow/src/state.rs:19]: I don't think this is a good idea. You are changing a boolean check by a function returning an error. And using the error enums to signal a condition that's not (not necessarily) an error.

File: escrow/tests/integration.rs [hunk 10]
```diff
@@ -86,155 +94,3 @@ fn proper_initialization() {
...
-    let msg = handle_res.messages.get(0).expect("no message");
-    assert_eq!(
-        msg,
-        &CosmosMsg::Bank(BankMsg::Send {
-            from_address: HumanAddr::from("cosmos2contract"),
-            to_address: HumanAddr::from("creator"),
-            amount: coins(1000, "earth"),
-        })
-    );
-}
```
Comment (maurolacy) [inline escrow/tests/integration.rs:240]: I think this is OK (removing integration tests, as discussed) but would have been better to do it as part of another PR. Or, at least, in a different (atomic) commit, so the intention / objective is clearer.

**Review by maurolacy — COMMENTED** [review]
lgtm overall.

File: escrow/src/contract.rs (hunk 3 above)
Comment (webmaster128) [reply escrow/src/contract.rs:23]: I agree. When a function starts with `is_`, this indicates a boolean return type.

**Review by maurolacy — APPROVED** [review]
LGTM.

File: escrow/src/error.rs [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+#[derive(Error, Debug)]
+pub enum ContractError {
+    #[error("{0}")]
+    Std(#[from] StdError),
+
+    #[error("unauthorized")]
+    Unauthorized {},
+
+    #[error("escrow expired (end_height {end_height})")]
+    Expired { end_height: u64, end_time: u64 },
```
Comment (webmaster128) [inline escrow/src/error.rs:13]: `end_height` and `end_time` of the escrow are both optional, as far as I know.

We can either wrap this on `Option` or add documention to the fields saying that e.g. `0` is used when unset.

File: escrow/src/contract.rs [hunk 12]
```diff
@@ -1,122 +1,133 @@
...
     };
+
     if state.is_expired(&env) {
-        Err(StdError::generic_err("creating expired escrow"))
-    } else {
-        config(&mut deps.storage).save(&state)?;
-        Ok(InitResponse::default())
+        return Err(ContractError::Expired {
+            end_height: env.block.height,
+            end_time: env.block.time,
```
Comment (webmaster128) [inline escrow/src/contract.rs:27]: This sets the current height and current time, not the end of the escrow. The values you look for are in `state`

File: escrow/src/contract.rs [hunk 13]
```diff
@@ -1,122 +1,133 @@
...
-            &deps.api.canonical_address(&env.contract.address)?,
-            &state.recipient,
-            amount,
-            "approve",
-        )
+    // throws error if state is expired
+    if state.is_expired(&env) {
+        return Err(ContractError::Expired {
+            end_height: env.block.height,
+            end_time: env.block.time,
```
Comment (webmaster128) [inline escrow/src/contract.rs:63]: As above, those are not the values of the escrow

File: escrow/src/contract.rs (hunk 12 above)
Comment (maurolacy) [reply escrow/src/contract.rs:27]: You're right!

File: escrow/src/contract.rs [hunk 14]
```diff
@@ -23,8 +23,8 @@ pub fn init<S: Storage, A: Api, Q: Querier>(
 
     if state.is_expired(&env) {
         return Err(ContractError::Expired {
-            end_height: env.block.height,
-            end_time: env.block.time,
+            end_height: msg.end_height,
+            end_time: msg.end_time,
```
Comment (maurolacy) [inline escrow/src/contract.rs:27]: You can use `state` here too for clarity, but it's the same.

File: escrow/src/error.rs [hunk 15]
```diff
@@ -9,8 +9,11 @@ pub enum ContractError {
     #[error("unauthorized")]
     Unauthorized {},
 
-    #[error("escrow expired (end_height {end_height})")]
-    Expired { end_height: u64, end_time: u64 },
+    #[error("escrow expired (end_height {end_height:?} end_time {end_time:?})")]
```
Comment (maurolacy) [inline escrow/src/error.rs]: I guess these will show up as `None` when unset. Which is a good alternative IMO.

**Review by maurolacy — APPROVED** [review]
This is correct now, AFAIK.

File: escrow/src/error.rs (hunk 15 above)
Comment (webmaster128) [reply escrow/src/error.rs]: Yes. For `Option` you need to use the `Debug` printing (thus the `:?`) because `Display` is not implemented

**Comment by webmaster128** [comment]
Thanks!
